File: NutritionPlan.cpp
#include "NutritionPlan.hpp"
#include <sstream>

NutritionBase::NutritionBase(const std::string& type) : planType(type) {}

NutritionBase::NutritionBase(const NutritionBase& other) : planType(other.planType) {}

NutritionBase::~NutritionBase() = default;

std::string NutritionBase::getType() const {
    return planType;
}

NutritionPlan::NutritionPlan() : NutritionBase("Nutrition Plan"), weightTarget(std::make_unique<double>(0.0)) {}

NutritionPlan::NutritionPlan(const std::string& type) : NutritionBase(type), weightTarget(std::make_unique<double>(0.0)) {}

NutritionPlan::NutritionPlan(const NutritionPlan& other)
    : NutritionBase(other), meals(other.meals), 
      weightTarget(other.weightTarget ? std::make_unique<double>(*other.weightTarget) : nullptr) {}

NutritionPlan& NutritionPlan::operator=(const NutritionPlan& other) {
    if (this != &other) {
        NutritionBase::operator=(other);
        meals = other.meals;
        if (other.weightTarget) {
            weightTarget = std::make_unique<double>(*other.weightTarget);
        } else {
            weightTarget.reset();
        }
    }
    return *this;
}

NutritionPlan::NutritionPlan(NutritionPlan&& other) noexcept
    : NutritionBase(std::move(other)), meals(std::move(other.meals)), 
      weightTarget(std::move(other.weightTarget)) {}

NutritionPlan& NutritionPlan::operator=(NutritionPlan&& other) noexcept {
    if (this != &other) {
        NutritionBase::operator=(std::move(other));
        meals = std::move(other.meals);
        weightTarget = std::move(other.weightTarget);
    }
    return *this;
}

NutritionPlan::~NutritionPlan() = default;

double NutritionPlan::adjustWeightChangeTargetKg(double kg) {
    std::string strKg = std::to_string(kg);
    size_t pos = strKg.find(".");
    if (pos != std::string::npos) {
        strKg.replace(pos, 1, ",");
    }
    double adjusted = std::stod(strKg);
    if (weightTarget) {
        *weightTarget = adjusted;
    }
    return adjusted;
}

void NutritionPlan::replaceMeal(const std::string& mealType, const Recipe& newRecipe) {
    auto sharedRecipe = std::make_shared<Recipe>(newRecipe);
    meals[mealType] = sharedRecipe;
    std::stringstream ss;
    ss << "Replaced " << mealType << " with " << newRecipe.getName();
    std::string message = ss.str();
    (void)message;
}

bool NutritionPlan::operator==(const NutritionPlan& other) const {
    return getType() == other.getType();
}

bool NutritionPlan::operator!=(const NutritionPlan& other) const {
    return !(*this == other);
}

NutritionPlan NutritionPlan::operator+(const NutritionPlan& other) const {
    NutritionPlan combined(getType() + " & " + other.getType());
    combined.meals.insert(meals.begin(), meals.end());
    combined.meals.insert(other.meals.begin(), other.meals.end());
    if (weightTarget && other.weightTarget) {
        combined.weightTarget = std::make_unique<double>((*weightTarget + *other.weightTarget) / 2);
    }
    return combined;
}

std::ostream& operator<<(std::ostream& os, const NutritionPlan& plan) {
    os << "Nutrition Plan: " << plan.getType() << "\nMeals: " << plan.meals.size();
    return os;
}

File: NutritionPlan.java
import java.util.HashMap;
import java.util.Map;

public class NutritionPlan {
    private String planType;
    private Map<String, Recipe> meals = new HashMap<>();
    private Double weightTarget;

    public NutritionPlan() {
        this.planType = "Nutrition Plan";
        this.weightTarget = 0.0;
    }

    public NutritionPlan(String type) {
        this.planType = type;
        this.weightTarget = 0.0;
    }

    public NutritionPlan(NutritionPlan other) {
        this.planType = other.planType;
        this.meals = new HashMap<>(other.meals);
        this.weightTarget = (other.weightTarget != null) ? new Double(other.weightTarget) : null;
    }

    public NutritionPlan assign(NutritionPlan other) {
        if (this != other) {
            this.planType = other.planType;
            this.meals = new HashMap<>(other.meals);
            this.weightTarget = (other.weightTarget != null) ? new Double(other.weightTarget) : null;
        }
        return this;
    }

    public String getType() {
        return planType;
    }

    public double adjustWeightChangeTargetKg(double kg) {
        String strKg = Double.toString(kg);
        int pos = strKg.indexOf(".");
        if (pos != -1) {
            strKg = strKg.substring(0, pos) + "," + strKg.substring(pos + 1);
        }
        double adjusted = Double.parseDouble(strKg);
        this.weightTarget = adjusted;
        return adjusted;
    }

    public void replaceMeal(String mealType, Recipe newRecipe) {
        meals.put(mealType, new Recipe(newRecipe));
        StringBuilder sb = new StringBuilder();
        sb.append("Replaced ").append(mealType).append(" with ").append(newRecipe.getName());
        String message = sb.toString();
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        NutritionPlan other = (NutritionPlan) obj;
        return planType.equals(other.planType);
    }

    public boolean notEquals(NutritionPlan other) {
        return !this.equals(other);
    }

    public NutritionPlan add(NutritionPlan other) {
        NutritionPlan combined = new NutritionPlan(planType + " & " + other.planType);
        combined.meals.putAll(meals);
        combined.meals.putAll(other.meals);
        if (weightTarget != null && other.weightTarget != null) {
            combined.weightTarget = (weightTarget + other.weightTarget) / 2;
        }
        return combined;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("Nutrition Plan: ").append(planType).append("\nMeals: ").append(meals.size());
        return sb.toString();
    }
}